core/src/main/java/hudson/util/FormFieldValidator.java

```diff
@@ -545,27 +545,21 @@ protected void check() throws IOException, ServletException {
                 if(path!=null) {
                     StringBuilder tokenizedPathBuilder = new StringBuilder();
                     for (String _dir : Util.tokenize(path.replace("\\", "\\\\"),File.pathSeparator)) {
-                        if (delimiter == null) {
-                          delimiter = ", ";
```

Comment:
After the change this assignments to `delimiter` is ignored, so the tokenized path is joined without delimiters. To maintain functionality, the whole `for` loop needs to be extracted instead of just the body.

Reply:
Thanks @zbynek. I've tried to do what you suggested.

Reply:
Having the iteratePath function is a nice improvement. I still think it would be more readable if iteratePath would return a File (or `Optional<File>` if you want to avoid `null`) and not require a Helper parameter. Since the next instruction after `helper.checkExecutable` is always return, this is equivalent. Then you don't need to fit the two checks into the same `Helper` interface, so you should be able to skip converting thrown exceptions to return values and back.

Reply:
Almost. Except this function does two things:
1. What you're talking about
2. helper.error("There's no such executable "+exe+" in PATH: "+tokenizedPath);

In order to handle the second case, the function would have to either return something that contains either the File or the error, or return a File or throw an error.

I wouldn't be terribly opposed to switching it to the latter. But, I'd definitely appreciate some guidance on it.

Reply:
Maybe you could make Helper an abstract class that has a result and an exception and utility methods hasResult() and hasException(). That way you could create the helper, provide it to iteratePath() and later return the appropriate information, which would avoid to create the arrays and use them to gather the results.

What do you think?

Reply:
>return something that contains either the File or the error,
Or you could use helper for reporting errors, but keep the signature of the of the function simple (return file or null). Either way please don't consider my comments as blockers.